features/perspectivepoller/src/test/java/org/opennms/netmgt/perspectivepoller/PerspectivePollerdIT.java

```diff
@@ -598,6 +615,46 @@ public void testPerspectivePollerThresholding() throws Exception {
         this.eventIpcManager.getEventAnticipator().verifyAnticipated();
     }
 
+    @Test
+    public void testStartWithEmptyApp() throws Exception {
+        this.databasePopulator.getTransactionTemplate().execute(tx -> {
+            final OnmsApplication app = new OnmsApplication();
+            app.setName("App Empty");
+            this.databasePopulator.getApplicationDao().save(app);
+
+            return null;
+        });
+
+        this.perspectivePollerd.start();
+    }
+
+    @Test
+    public void testStartWithDuplicatedService() throws Exception {
```

Comment:
Just to be sure to cover all cases - can you add additional tests for:
* application with services and without locations
* application with locations and without services

Reply:
Good point. Done.